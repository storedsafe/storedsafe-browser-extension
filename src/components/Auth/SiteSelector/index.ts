import { SiteSelector } from './SiteSelector';

export default SiteSelector;
