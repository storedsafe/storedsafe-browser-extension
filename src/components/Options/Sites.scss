@import '../../variables';

.sites {
  .sites-article {
    .sites-article-header {
      padding: $spacing $spacing * 2;
      background-color: $color-primary-1;
      color: $color-primary-0;
    }
  }

  .sites-list {
    > li {
      list-style: none;
      display: flex;
      flex-direction: row;
      align-items: space-between;
      justify-content: space-between;
      padding-left: $spacing * 2;
    }

    > li:nth-child(odd) {
      background-color: $color-bg-hover;
    }

    span {
      padding: $spacing;
    }
  }

  .sites-add {
    .sites-add-form {
      display: flex;
      flex-direction: column;
      align-items: stretch;
      width: 100%;
      padding: $spacing;
    }

    label:not(:last-child),
    .button:not(:last-child),
    .message:not(:last-child) {
      margin-bottom: $spacing;
    }

    input {
      width: 100%;
    }
  }
}

@media (min-width: 600px) {
  .sites {
    .sites-article {
      .sites-article-header {
        min-width: max-content;
      }
    }

    .sites-add {
      display: flex;
      flex-direction: row;
    }
  }
}

