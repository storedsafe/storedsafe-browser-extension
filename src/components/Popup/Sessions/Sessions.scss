@import '../../../variables';

.sessions {
  display: flex;
  flex-direction: column;

  &.selected {
    height: 100%;
    flex-grow: 1;
  }

  .sessions-menu {
    display: none;
  }

  &.selected .sessions-menu {
    flex-grow: 1;
    display: flex;
    flex-direction: column;

    .session-item {
      display: flex;
      flex-direction: column;
      overflow: auto;
    }

    .session-item-button {
      padding: $spacing;
      cursor: pointer;
      outline: 0;
      border: 0;
      background-color: $color-bg;
      display: flex;
      flex-direction: row;
      align-items: center;
      justify-content: space-between;
    }
  }

  .sessions-toggle {
    cursor: pointer;
    user-select: none;
    box-shadow: 0 -1px 1px 0 rgba(0, 0, 0, 0.1);
  }

  .sessions-status {
    background-color: $color-primary-0;
    display: flex;
    flex-direction: row;
    align-items: center;
    padding: $spacing;
    width: 100%;

    span {
      color: $color-primary-1;
      font-size: 1.1em;
      font-weight: bold;
      flex-grow: 1;
      text-align: center;
    }
  }

  .sessions-open {
    background-color: $color-primary-1;
    color: $color-primary-0;
    text-align: center;
    padding: $spacing / 2;
  }
}
