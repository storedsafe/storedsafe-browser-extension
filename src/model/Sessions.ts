export interface Session {
  apikey: string;
  token: string;
  createdAt: number;
  lastActive: number;
}

/**
 * Sessions object.
 * */
export interface Sessions {
  [url: string]: Session;
}

/**
 * Get sessions from local storage.
 * @return Sessions Promise containing Sessions object.
 * */
export const get = (): Promise<Sessions> => (
  browser.storage.local.get('sessions')
  .then(({ sessions }) => {
    if (sessions) {
      return sessions || {};
    }
    return {};
  })
)

/**
 * Commit Sessions object to sync storage.
 * @param sessions New Sessions object.
 * */
export const set = (sessions: Sessions): Promise<void> => {
  return browser.storage.local.set({ sessions });
}
