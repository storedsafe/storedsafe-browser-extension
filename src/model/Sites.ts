const systemStorage = browser.storage.managed;
const userStorage = browser.storage.sync;

export interface Site {
  url: string;
  apikey: string;
}

export interface SiteCollection {
  system: Site[];
  user: Site[];
}

/**
 * Layout in storage.
 * */
interface Sites {
  sites: Site[];
}

/**
 * Get sites from system and user storage.
 * @return Sites Promise containg SiteCollection object.
 * */
export const get = (): Promise<SiteCollection> => {
  return Promise.all<Sites, Sites>([
    systemStorage.get('sites') as Promise<Sites>,
    userStorage.get('sites') as Promise<Sites>,
  ]).then(([{ sites: system }, { sites: user }]) => {
    return { system: system || [], user: user || [] };
  });
}

/**
 * Commit Sites object to user storage (managed sites are ignored).
 * @param sites New Sites object.
 * */
export const set = (sites: SiteCollection): Promise<void> => {
  return userStorage.set({ sites: sites.user });
}
