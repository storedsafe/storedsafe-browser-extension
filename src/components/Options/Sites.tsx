import React, { useState } from 'react';
import { useStorage } from '../../hooks/useStorage';
import { Button } from '../UI';
import { useForm } from '../../hooks/useForm';
import './Sites.scss';

/**
 * Form values in the add site form.
 * */
interface AddSiteValues {
  url: string;
  apikey: string;
}

export const Sites: React.FunctionComponent = () => {
  const [loading, setLoading] = useState<{ [key: string]: boolean }>({});

  /**
   * Initial values for the add site form.
   * */
  const addSiteInitialValues = { url: '', apikey: '' };
  const [
    addSiteValues, events, reset
  ] = useForm<AddSiteValues>(addSiteInitialValues);
  const { state, dispatch, isInitialized } = useStorage();

  if (!isInitialized) return <p>Loading...</p>;

  const { system, user } = state.sites.collections;

  /**
   * List of sites managed by the system administrator
   * */
  const managedSites = system.map((site) => (
    <li key={site.url}><span>{site.url}</span></li>
  ));

  /**
   * Creates a closure around a url to be used as a callback
   * funtion for a button click to remove a site from storage.
   * */
  const onRemove = (
    removeUrl: string
  ): () => void => (): void => {
    setLoading({ ...loading, [removeUrl]: true });
    const id = user.findIndex(({ url }) => url === removeUrl);
    dispatch({
      sites: {
        type: 'remove', id
      },
    }, {
      onSuccess: () => {
        setLoading({ ...loading, [removeUrl]: false })
      },
      onError: () => {
        setLoading({ ...loading, [removeUrl]: false })
      },
    })
  };

  /**
   * List of sites managed by the user.
   * */
  const userSites = user.map((site) => (
    <li key={site.url}>
      <span>{site.url}</span>
      <Button
        color="danger"
        onClick={onRemove(site.url)}
        isLoading={loading[site.url] || false}>
        Delete
      </Button>
    </li>
  ));

  /**
   * Callback function to add site after the add site form
   * is submitted.
   * */
  const onAdd = (
    event: React.FormEvent<HTMLFormElement>
  ): void => {
    setLoading({ ...loading, add: true });
    event.preventDefault();
    const { url, apikey } = addSiteValues;
    dispatch({
      sites: {
        type: 'add',
        site: { url, apikey },
      },
    }, {
      onSuccess: () => {
        setLoading({ ...loading, add: false });
        reset();
      },
      onError: (error) => {
        setLoading({ ...loading, add: false })
        console.error(error);
      },
    })
  }

  return (
    <section className="sites">
      <article className="sites-article sites-add">
        <header className="sites-article-header">
          <h3>Add new site</h3>
        </header>
        <form className="sites-add-form" onSubmit={onAdd}>
          <label htmlFor="url">
            URL
            <input
              type="text"
              name="url"
              id="url"
              placeholder="URL"
              required
              value={addSiteValues.url}
              {...events}
            />
          </label>
          <label htmlFor="apikey">
            API Key
            <input
              type="text"
              name="apikey"
              id="apikey"
              placeholder="API Key"
              required
              value={addSiteValues.apikey}
              {...events}
            />
          </label>
          <Button
            color="accent"
            type="submit"
            isLoading={loading.add || false}>
            Add Site
          </Button>
        </form>
      </article>

      {userSites.length > 0 && (<article
        className="sites-article sites-user">
        <header className="sites-article-header">
          <h3>Sites managed by user</h3>
        </header>
        <ul className="sites-list sites-user-list">
          {userSites}
        </ul>
      </article>)
      }

      {managedSites.length > 0 && (<article
        className="sites-article sites-managed">
        <header className="sites-article-header">
          <h3>Sites managed by organization</h3>
        </header>
        <ul className="sites-list sites-managed-list">
          {managedSites}
        </ul>
      </article>)
      }

    </section>
  );
};
