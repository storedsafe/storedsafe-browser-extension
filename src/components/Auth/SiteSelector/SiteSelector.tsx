import * as React from 'react';
import { useStorage } from '../../../state/StorageState';
import { Field, Option, OnChangeCallback } from '../../Form';
import { Site } from '../../../model/Sites';
import { Sessions } from '../../../model/Sessions';

export interface SiteSelectorProps {
  sites: Site[];
  sessions: Sessions;
  value: number;
}
export const SiteSelector: React.SFC = () => {
  const [state, dispatch] = useStorage();

  const sites = state.sites.list;
  const sessions = state.sessions;
  const selected = state.authState.selected;
  const onChange: OnChangeCallback = (url: string) => {
    dispatch({ authState: { type: 'setSelected', url } });
  };

  const options: Option[] = sites.map((site) => ({
    title: `${sessions[site.url] !== undefined ? '[*] ' : ''}${site.url}`,
    value: site.url,
  }));

  return (
    <Field
      type="select"
      name="site"
      label="Site"
      value={selected}
      onChange={onChange}
      options={options} />
  );
};
