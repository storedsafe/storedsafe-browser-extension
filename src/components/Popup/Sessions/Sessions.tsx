import React, { useState } from 'react';
import { useStorage } from '../../../state/StorageState';
import { Site } from '../../../model/Sites';
import { Session } from '../../../model/Sessions';
import { OnlineIndicator } from '../../Layout';
import Auth from '../Auth';
import './Sessions.scss';

interface SessionItemProps {
  site: Site;
  session?: Session;
}

const SessionItem: React.FunctionComponent<SessionItemProps> = ({
  site,
  session,
}: SessionItemProps) => {
  const [showAuth, setShowAuth] = useState<boolean>(false);
  const { url } = site;
  const isOnline = session !== undefined;
  const minutesActive = isOnline && Math.floor((Date.now() - session.createdAt) / (1000 * 60));

  return (
    <article className="session-item">
      <button type="button" className="session-item-button" onClick={(): void => setShowAuth(!showAuth)}>
        <span>
          {url}
          {isOnline && minutesActive !== undefined && ` (${minutesActive} min)` || ''}
        </span>
        <OnlineIndicator online={isOnline} />
      </button>
      {showAuth && <Auth url={url} />}
    </article>
  );
};

interface SessionsProps {
  selected: boolean;
  toggleActive: () => void;
}

export const Sessions: React.FunctionComponent<SessionsProps> = ({
  selected,
  toggleActive,
}: SessionsProps) => {
  const [state] = useStorage();

  const numberOfSessions = Object.keys(state.sessions).length;
  let sessionsText = 'No active sessions';

  if (numberOfSessions > 0) {
    sessionsText = `${numberOfSessions} active session${numberOfSessions === 1 ? '' : 's'}`;
  }

  return (
    <section className={`sessions${selected ? ' selected' : ''}`}>
      <article className="sessions-menu">
        {state.sites.list.map((site) => (
          <SessionItem
            key={site.url}
            site={site}
            session={state.sessions[site.url]}
          />
        ))}
      </article>
      <article
        className="sessions-toggle"
        onClick={toggleActive}>
        <div className="sessions-open">
          <span>show/hide sessions</span>
        </div>
        <div className="sessions-status">
          <span>{sessionsText}</span>
          <OnlineIndicator online={numberOfSessions > 0} />
        </div>
      </article>
    </section>
  );
};
