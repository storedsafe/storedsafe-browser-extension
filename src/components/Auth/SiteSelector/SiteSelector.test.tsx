import * as React from 'react';
import { shallow } from 'enzyme';
import { Site } from '../../../model/Sites';
import { Sessions } from '../../../model/Sessions';
import { SiteSelector } from './SiteSelector';

const sites: Site[] = [
  { url: 'mock1.example.com', apikey: 'mockApikey' },
  { url: 'mock2.example.com', apikey: 'mockApikey' },
];
const sessions: Sessions = {
  [sites[0].url]: {
    apikey: sites[0].apikey,
    token: 'mockToken',
    createdAt: 0,
    lastActive: 0,
  },
};

test('<SiteSelector />', () => { const wrapper = shallow(<SiteSelector />);
  expect(wrapper).toMatchSnapshot();
});
