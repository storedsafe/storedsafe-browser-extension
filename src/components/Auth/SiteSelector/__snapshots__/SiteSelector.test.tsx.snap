// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<SiteSelector /> 1`] = `ShallowWrapper {}`;
